include/crab/domains/object_domain.hpp

```diff
@@ -0,0 +1,1303 @@
+    for (auto &kv : right.m_flds_rep_map) {
+      const variable_t &v = kv.first;
+      auto it = out_flds_rep_map.find(v);
+      if (it == out_flds_rep_map.end()) {
+        out_flds_rep_map.insert({v, kv.second});
+      }
+    }
+
+    object_domain_t res(std::move(out_base_dom), std::move(out_odi_map),
+                        std::move(out_obj_info_env),
+                        std::move(out_flds_rep_map));
+    return res;
+  }
+
+  // compare two abstract states, return boolean
+  bool less_than_eq(const object_domain_t &left,
+                    const object_domain_t &right) const {
+
+    if (!(left.m_obj_info_env <= right.m_obj_info_env)) {
+      CRAB_LOG("object-leq", crab::outs() << "Result3=0\n";);
+      return false;
+    }
+
+    bool res = left.m_base_dom <= right.m_base_dom;
+    CRAB_LOG("object-leq", crab::outs() << "Result4=" << res << "\n";);
+
+    res = left.m_odi_map <= right.m_odi_map;
+
+    CRAB_LOG("object-leq", crab::outs() << "Result5=" << res << "\n";);
+    return res;
+  }
+
+  static void ERROR_IF_NOT_REGION(const variable_t &v, unsigned line) {
+    if (!v.get_type().is_region()) {
+      CRAB_ERROR(v, ":", v.get_type(), " is not a region at line ", line);
+    }
+  }
+
+  static void ERROR_IF_ARRAY_REGION(const variable_t &v, unsigned line) {
+    if (v.get_type().is_array_region()) {
+      CRAB_ERROR(v, ":", v.get_type(), " cannot contain an array at line ",
+                 line);
+    }
+  }
+
+  static void ERROR_IF_NOT_REF(const variable_t &v, unsigned line) {
+    if (!v.get_type().is_reference()) {
+      CRAB_ERROR(v, ":", v.get_type(), " is not a reference at line ", line);
+    }
+  }
+
+  static void ERROR_IF_NOT_INT(const variable_t &v, unsigned line) {
+    if (!v.get_type().is_integer()) {
+      CRAB_ERROR(v, ":", v.get_type(), " is not an integer at line ", line);
+    }
+  }
+
+  variable_t get_object_rep_by_rgn(variable_t rgn) {
+    auto it = m_flds_rep_map.find(rgn);
+    if (it == m_flds_rep_map.end()) {
+      CRAB_ERROR(
+          domain_name(),
+          "::get_object_rep_by_rgn, the odi map does not include the region ",
+          rgn, " belonging to an abstract object");
+    }
+    return it->second;
+  }
+
+  void get_rgn_vec(variable_t rep, variable_vector_t &rgn_vec) {
+    for (auto kv : m_flds_rep_map) {
+      if (kv.second == rep) {
+        rgn_vec.push_back(kv.second);
+      }
+    }
+  }
+
+  template <class DOM>
+  void assign_interval(DOM &dom, const variable_t v, interval_t ival) {
+    dom -= v;
+    boost::optional<number_t> lb = ival.lb().number();
+    boost::optional<number_t> ub = ival.ub().number();
+    if (lb) {
+      dom += (*lb <= v);
+    }
+    if (ub) {
+      dom += (v <= *ub);
+    }
+  }
+
+  void move_singleton_to_odi_map(variable_t rep) {
+    variable_vector_t rgn_vec;
+    base_abstract_domain_t tmp_base(m_base_dom);
+    tmp_base.project(rgn_vec);
+    // Be careful of the following operation if base dom and object dom are
```

Comment:
you can use a `static_assert` to check for what the comment is saying, or implement a cast operation